### PR TITLE
[WIP] Add Swin Transformer

### PR DESCRIPTION
**Pull request by zeliu98**

### CODE REVIEW

File: mmcv_custom/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
+# Copyright (c) Open-MMLab. All rights reserved.
```
Comment (xvjiarui) [inline mmcv_custom/checkpoint.py:1]: We may remove this file and directory.

File: mmseg/models/backbones/swin_transformer.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+                qk_scale=qk_scale,
+                drop_rate=drop_rate,
+                attn_drop_rate=attn_drop_rate,
+                drop_path_rate=dpr[:depths[i]],
+                downsample=downsample,
+                act_cfg=act_cfg,
+                norm_cfg=norm_cfg,
+                auto_pad=auto_pad,
+                init_cfg=None,
+            )
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:627]:
```suggestion
                init_cfg=None)
```

File: mmseg/models/backbones/swin_transformer.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        if stride is None:
+            stride = kernel_size
+        kernel_size = to_2tuple(kernel_size)
+        stride = to_2tuple(stride)
+        padding = to_2tuple(padding)
+        dilation = to_2tuple(dilation)
+        self.sampler = nn.Unfold(kernel_size, dilation, padding, stride)
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:65]: Check if https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/87e6f90577435c94f3e92c7db1d36edc234d91f6/mmseg/models/backbones/swin_transformer.py#L283-L286 is also correctly handled.

File: mmseg/models/backbones/swin_transformer.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+    @staticmethod
+    def double_step_seq(step1, len1, step2, len2):
+        seq1 = torch.arange(0, step1 * len1, step1)
+        seq2 = torch.arange(0, step2 * len2, step2)
+        return (seq1[:, None] + seq2[None, :]).reshape(1, -1)
+
+
+@ATTENTION.register_module()
+class ShiftWindowMSA(BaseModule):
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:216]: Missing docstring.

File: mmseg/models/backbones/swin_transformer.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+
+    def __init__(self,
+                 input_resolution,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:219]: In our setting, the input size may change during infer or training. We should determine input size when initializing the module.

File: mmseg/models/backbones/swin_transformer.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+        B, H, W, C = x.shape
+        window_size = self.window_size
+        x = x.view(B, H // window_size, window_size, W // window_size,
+                   window_size, C)
+        windows = x.permute(0, 1, 3, 2, 4, 5).contiguous()
+        windows = windows.view(-1, window_size, window_size, C)
+        return windows
+
+
+class SwinBlock(BaseModule):
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:366]: Missing docstring.

File: mmseg/models/backbones/swin_transformer.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+
+    def __init__(self,
+                 input_resolution,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:369]: Similarly, input_size should be unknown.

File: mmseg/models/backbones/swin_transformer.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+        else:
+            self.downsample = None
+
+    def forward(self, query):
+        for block in self.blocks:
+            query = block(query)
+
+        if self.downsample:
+            query = self.downsample(query)
+        return query
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin_transformer.py:494]: We should have something like 
https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/87e6f90577435c94f3e92c7db1d36edc234d91f6/mmseg/models/backbones/swin_transformer.py#L362

Comment (xvjiarui) [reply mmseg/models/backbones/swin_transformer.py:494]: I suggest pack (H, W) into `hw_shape`, and forward it also.

File: mmseg/models/backbones/swin_transformer.py (hunk 3 above)
Comment (sennnnn) [reply mmseg/models/backbones/swin_transformer.py:65]: The padding may need to calculate by users.

File: mmseg/models/backbones/swin_transformer.py (hunk 8 above)
Comment (sennnnn) [reply mmseg/models/backbones/swin_transformer.py:494]: Actually, the H, W are wrapped in the attributes of PatchMerging. H, W = self.output_resolution.

File: mmseg/models/backbones/swin.py [hunk 9]
```diff
@@ -0,0 +1,754 @@
...
+        seq1 = torch.arange(0, step1 * len1, step1)
+        seq2 = torch.arange(0, step2 * len2, step2)
+        return (seq1[:, None] + seq2[None, :]).reshape(1, -1)
+
+
+@ATTENTION.register_module()
+class ShiftWindowMSA(BaseModule):
+
+    def __init__(self,
+                 input_resolution,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:225]: We shouldn't have `input_resolution` arg.
Reference: https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/87e6f90577435c94f3e92c7db1d36edc234d91f6/mmseg/models/backbones/swin_transformer.py#L156

File: mmseg/models/utils/embed.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
+import torch
```
Comment (xvjiarui) [inline mmseg/models/utils/embed.py:1]: Modified from xxx.

File: mmseg/models/backbones/swin.py [hunk 11]
```diff
@@ -0,0 +1,717 @@
...
+                kernel_size=kernel_size,
+                stride=stride,
+                padding=padding,
+                dilation=dilation,
+                norm_cfg=norm_cfg,
+                init_cfg=None)
+        else:
+            self.downsample = None
+
+    def forward(self, x, H, W):
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:459]: We may pack `H, W` into `hw_shape` as a tuple. 
Reference https://github.com/facebookresearch/SlowFast/blob/2090f2918ac1ce890fdacd8fda2e590a46d5c734/slowfast/models/video_model_builder.py#L1002

File: mmseg/models/backbones/swin.py [hunk 12]
```diff
@@ -0,0 +1,717 @@
...
+            block.W = W
+            x = block(x)
+
+        if self.downsample:
+            stage_out = x
+            x = self.downsample(x, H, W)
+            DH, DW = (H + 1) // 2, (W + 1) // 2
+            return stage_out, H, W, x, DH, DW
+        else:
+            return x, H, W, x, H, W
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:471]: The output should be `x, hw_shape`. 
We don't need the previous `hw_shape`.

File: mmseg/models/backbones/swin.py (hunk 11 above)
Comment (xvjiarui) [reply mmseg/models/backbones/swin.py:459]: We also need a docstring for this.

File: mmseg/models/backbones/swin.py [hunk 13]
```diff
@@ -0,0 +1,717 @@
...
+                dilation=dilation,
+                norm_cfg=norm_cfg,
+                init_cfg=None)
+        else:
+            self.downsample = None
+
+    def forward(self, x, H, W):
+        for block in self.blocks:
+            block.H = H
+            block.W = W
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:462]: Setting attributes from outside is not recommended. We may pass `hw_shape`.

File: mmseg/models/backbones/swin.py [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+        x = x.view(B, H // window_size, window_size, W // window_size,
+                   window_size, C)
+        windows = x.permute(0, 1, 3, 2, 4, 5).contiguous()
+        windows = windows.view(-1, window_size, window_size, C)
+        return windows
+
+
+class SwinBlock(BaseModule):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:341]: Missing docstring.

File: mmseg/models/backbones/swin.py [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+
+        if pad_r > 0 or pad_b:
+            x = x[:, :H, :W, :].contiguous()
+
+        x = x.view(B, H * W, C)
+
+        x = self.drop(x)
+        return x
+
+    def window_reverse(self, windows, H, W):
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:321]: Missing docstring.

File: mmseg/models/backbones/swin.py [hunk 16]
```diff
@@ -0,0 +1,720 @@
...
+        window_size = self.window_size
+        B = int(windows.shape[0] / (H * W / window_size / window_size))
+        x = windows.view(B, H // window_size, W // window_size, window_size,
+                         window_size, -1)
+        x = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(B, H, W, -1)
+        return x
+
+    def window_partition(self, x):
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:329]: Missing docstring.

File: mmseg/models/backbones/swin.py [hunk 17]
```diff
@@ -0,0 +1,720 @@
...
+    def double_step_seq(step1, len1, step2, len2):
+        seq1 = torch.arange(0, step1 * len1, step1)
+        seq2 = torch.arange(0, step2 * len2, step2)
+        return (seq1[:, None] + seq2[None, :]).reshape(1, -1)
+
+
+@ATTENTION.register_module()
+class ShiftWindowMSA(BaseModule):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:216]: Missing docstring.

**Review by xvjiarui — COMMENTED** [review]
We could refer to the original implementation for the missing docstring.

File: mmseg/models/backbones/swin.py [hunk 18]
```diff
@@ -0,0 +1,714 @@
...
+            Defaults: None.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size=2,
+                 stride=None,
+                 padding=0,
+                 dilation=1,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:53]: Remove these args. Make stride=kernel_size.

File: mmseg/models/backbones/swin.py [hunk 19]
```diff
@@ -0,0 +1,821 @@
...
+        norm_cfg (dict, optional): Config dict for normalization layer.
+            Defaults: dict(type='LN').
+        init_cfg (dict, optional): The extra config for initialization.
+            Defaults: None.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size=2,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:50]: We may remove this.

File: mmseg/models/backbones/swin.py [hunk 20]
```diff
@@ -0,0 +1,821 @@
...
+    def __init__(self,
+                 pretrain_img_size=224,
+                 in_channels=3,
+                 embed_dims=96,
+                 patch_size=4,
+                 window_size=7,
+                 mlp_ratio=4,
+                 depths=(2, 2, 6, 2),
+                 num_heads=(3, 6, 12, 24),
+                 strides=(None, None, None, None),
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:620]: Why the default is none?

File: mmseg/models/backbones/swin.py [hunk 21]
```diff
@@ -0,0 +1,821 @@
...
+
+class SwinBlockSequence(BaseModule):
+    """Implements one stage in Swin Transformer.
+
+    Args:
+        embed_dims (int): The feature dimension.
+        num_heads (int): Parallel attention heads.
+        feedforward_channels (int): The hidden dimension for FFNs.
+        depth (int): The number of blocks in this stage.
+        kernel_size (int): The kernel_size of patch merging.
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:467]: We may also remove this. Use `stride` only.

File: mmseg/models/backbones/swin.py [hunk 22]
```diff
@@ -0,0 +1,821 @@
...
+        stride (int): The kernel slide stride of patch merging.
+        padding (int): The padding length of patch merging.
+        dilation (int): The dilation rate of kernel of patch merging.
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:470]: Not needed.

File: mmseg/models/backbones/swin.py [hunk 23]
```diff
@@ -0,0 +1,821 @@
...
+
+    def __init__(self,
+                 embed_dims,
+                 num_heads,
+                 feedforward_channels,
+                 depth,
+                 kernel_size,
+                 stride,
+                 padding,
+                 dilation,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:496]: Keep stride only.

File: mmseg/models/backbones/swin.py [hunk 24]
```diff
@@ -0,0 +1,821 @@
...
+                stride=stride,
+                padding=padding,
+                dilation=dilation,
+                window_size=window_size,
+                qkv_bias=qkv_bias,
+                qk_scale=qk_scale,
+                drop_rate=drop_rate,
+                attn_drop_rate=attn_drop_rate,
+                drop_path_rate=dpr[:depths[i]],
+                downsample=downsample,
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:717]: How about constructing `downsample` first, and pass the module

File: mmseg/models/backbones/swin.py [hunk 25]
```diff
@@ -0,0 +1,786 @@
...
+        strides (tuple[int] | tuple[None], optional): The patch merging or
+            patch embedding stride of each Swin Transformer stage. (When stride
+            is set to None, stride will be equal to kernel size.)
+            Default: (None, None, None, None).
+        paddings (tuple[int], optional): The patch merging or patch
+            embedding padding length of each Swin Transformer stage.
+            Default: (0, 0, 0, 0).
+        dilations (tuple[int], optional): The patch merging or patch
+            embedding kernel dilation rate of each Swin Transformer stage.
+            Default: (1, 1, 1, 1).
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:558]: These are no longer needed.

File: mmseg/models/backbones/swin.py [hunk 26]
```diff
@@ -0,0 +1,786 @@
...
+                drop_path_rate=dpr[:depths[i]],
+                downsample=downsample,
+                act_cfg=act_cfg,
+                norm_cfg=norm_cfg,
+                init_cfg=None)
+            self.stages.append(stage)
+
+            dpr = dpr[depths[i]:]
+            if downsample:
+                in_channels = in_channels * 2
```
Comment (xvjiarui) [inline mmseg/models/backbones/swin.py:690]: Move to the definition of `downample`.

**Review by xvjiarui — APPROVED** [review]
